src/btree/bt_random.c

```diff
@@ -365,14 +365,22 @@ __wt_random_descent(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags, WT_
 
     btree = S2BT(session);
     current = NULL;
-    retry = 100;
+    retry = 0;
     /*
      * This function is called by eviction to find a random page in the cache. That case is
      * indicated by the WT_READ_CACHE flag. Ordinary lookups in a tree will read pages into cache as
      * needed.
      */
     eviction = LF_ISSET(WT_READ_CACHE);
 
+    if (eviction) {
+        if (F_ISSET(S2C(session)->cache, WT_CACHE_EVICT_RANDOM_RETRY)) {
+            WT_INTL_INDEX_GET(session, (&btree->root)->page, pindex);
+            retry = pindex->entries;
```

Comment:
I've seen cases where there are 10k entries on an internal page. Does the logic hold in that case? Should this have an upper bound? I can imagine a giant tree that is only being appended to would have a single page in the cache, but a lot of on-disk refs. I'm not sure what the odss of finding that single ref are - but repeatedly doing a random position until you find it doesn't feel like the right algorithm.

Reply:
Please correct me if I am wrong but 10,000 entries is the maximum for an internal page before a split is triggered.

I understand that in such cases, the function can become costly, and even then, there's only about a 60% chance that we'll find the entry in-memory. 

If we've reached this point in the code, it indicates that this tree is contributing enough to the cache for either clean, dirty or update eviction. For clean eviction, this should be acceptable, as we may only end up in this situation rarely. 

However, for dirty and update evictions, especially in an append-only workload, this could be a expensive. However, this cost is justified as it's better to pay this expense now than to face greater costs in the future by frequently walking through the same pages.

If an effort is made in the random function to find the right entry and avoiding returning null position (meaning we start from the begging or the end). We may anyway repeatedly walk 10000 pages in the tree, and we will be doing this in each eviction pass.

Reply:
I am happy to set an upper cap if that seems safer.

Reply:
If I understand correctly, the goal is to return a valid ref from `__wt_random_descent` more often than we do today. 

This makes me think we should always retry at least 100 times -- today's limit. Otherwise we might have cases where we have, say 1 cached ref out of 50, and we get unlucky in 50 tries, but we might have found that ref with 100 tries. I.e., this change could do worse than our current implementation.

If we do that, then any higher setting for retry when evicting will be strictly better than the behavior we have today. The question, it seems to me, is how much better we want/need to do.

It sounds like a worst case might be a node with 10K entries, only one of which is in cache. `retry = 1000` would give us only a 10% chance of finding that cached ref. `retry = 10000` raises the odds to 73%. Neither of those numbers are great. But they are both better than today. So if that's good enough, we can just decide on a maximum -- whether it's a constant or based on the number of entries.

And it may be that if we get to a good enough number (10%), then eviction will visit a tree often enough that we will eventually find stuff by doing this random probing multiple times?

If we want to be certain of doing better then we need to either be willing to keep picking random numbers (100K tries will find 1 in 10K entries 99.99% of the time) or we need a different selection algorithm.

Reply:
Present retry limit is 0. The limit 100 is for random search not eviction.

Understood, I think the cap 10000 makes sense to me. As I have seen cases where it may take a while to split root page and the entries may grow larger than 10000 @agorrod?

Reply:
I have done the same in [8435e5c](https://github.com/wiredtiger/wiredtiger/pull/10437/commits/8435e5c7dcd04799722d131c8ad7d69adcbd9262). Let me know what you think, thanks!